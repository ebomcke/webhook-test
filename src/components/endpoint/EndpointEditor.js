import React, { Component } from 'react';
import { compose } from 'recompose';
import { withFirestore } from 'react-firestore';
import { withFirebase } from '../../contexts/Firebase';
import { withAccount } from '../../contexts/Account';
import { withNavigation } from '../../contexts/Navigation';

class EndpointEditor extends Component {
  constructor(props) {
    super(props);
    let { endpoint } = this.props;
    if (!endpoint) {
      endpoint = this.generateNewEndpoint();
    }
    this.state = {
      id: endpoint.id,
      path: endpoint.path,
      defaultResponseBody: endpoint.defaultResponse.body,
      defaultResponseStatusCode: endpoint.defaultResponse.statusCode,
    };
  }

  generateNewEndpoint = () => ({
    path: '',
    defaultResponse: {
      statusCode: 200,
      body: JSON.stringify(
        {
          status: 'OK',
        },
        null,
        2,
      ),
    },
  });

  onStatusCodeChange = (statusCode) => {
    this.setState({
      defaultResponseStatusCode: statusCode,
    });
  }

  onUrlChange = (url) => {
    url = url
      .toLowerCase()
      .replace(/ /gi, '')
      .replace(/^\//, '');
    this.setState({
      path: url,
    });
  }

  onResponseBodyChange = (value) => {
    try {
      JSON.parse(value);
      this.setState({
        defaultResponseBody: value,
      });
    } catch (e) {}
  }

  save = async () => {
    const { firebase, firestore, account, navigate } = this.props;
    let endpoint = {
      account: firestore.collection('accounts').doc(account.id),
      path: this.state.path,
      lastActive: firebase.firestore.Timestamp.now(),
      defaultResponse: {
        statusCode: this.state.defaultResponseStatusCode,
        body: this.state.defaultResponseBody,
      },
    };
    if (this.state.id) {
      endpoint.id = this.state.id;
    }
    if (endpoint.id) {
      await firestore
        .collection('endpoints')
        .doc(endpoint.id)
        .set(endpoint);
    } else {
      await firestore.collection('endpoints').add(endpoint);
    }
    navigate({
      name: 'Endpoints',
    });
  }

  render() {
    const { children } = this.props;
    return children({
      endpoint: this.state,
      onResponseBodyChange: this.onResponseBodyChange,
      onUrlChange: this.onUrlChange,
      onStatusCodeChange: this.onStatusCodeChange,
      save: this.save,
    });
  }
}

export default compose(
  withFirebase,
  withFirestore,
  withNavigation,
  withAccount,
)(EndpointEditor);
