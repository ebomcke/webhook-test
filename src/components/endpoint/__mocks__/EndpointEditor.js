const mockEndpointEditor = {
    endpoint: {
        path: 'unit-test-path',
        defaultResponse: {
            statusCode: 200,
            body: JSON.stringify({status:'OK'})
        }
    },
    onResponseBodyChange: jest.fn(),
    onUrlChange: jest.fn(),
    onStatusCodeChange: jest.fn(),
    save: jest.fn(() => Promise.resolve())
}

export { mockEndpointEditor };