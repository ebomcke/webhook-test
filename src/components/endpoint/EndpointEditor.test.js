import React from 'react';
import EndpointEditor from './EndpointEditor';
import renderer from 'react-test-renderer';

describe('Rendering tests', () => {
  test('New endpoint', () => {
    const endpointEditor = renderer.create(
      <EndpointEditor>
        {endpointEditor => <div endpointEditor={endpointEditor} />}
      </EndpointEditor>,
    );
    let tree = endpointEditor.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Existing endpoint', () => {
    const endpointEditor = renderer.create(
      <EndpointEditor
        endpoint={{
          id: 'existing-endpoint',
          path: 'existing-endpoint-path',
          defaultResponse: {
            statusCode: 418,
            body: JSON.stringify({ unitTest: 'true' }),
          },
        }}
      >
        {endpointEditor => <div endpointEditor={endpointEditor} />}
      </EndpointEditor>,
    );
    let tree = endpointEditor.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Interactions tests', () => {
  let endpointEditor;

  beforeEach(() => {
    endpointEditor = renderer.create(
        <EndpointEditor>
          {endpointEditor => (
            <div hasEditor={true} endpointEditor={endpointEditor} />
          )}
        </EndpointEditor>,
      );
  })

  test('onStatusCodeChange updates state value', () => {
    endpointEditor.root
      .findByProps({ hasEditor: true })
      .props.endpointEditor.onStatusCodeChange(418);
    let tree = endpointEditor.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('onUrlChange updates state value', () => {
    endpointEditor.root
      .findByProps({ hasEditor: true })
      .props.endpointEditor.onUrlChange('new-url-value');
    let tree = endpointEditor.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('onUrlChange validates url before updating state', () => {
    endpointEditor.root
      .findByProps({ hasEditor: true })
      .props.endpointEditor.onUrlChange('/NEW url Value');
    let tree = endpointEditor.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('onResponseBodyChange updates state value', () => {
    endpointEditor.root
      .findByProps({ hasEditor: true })
      .props.endpointEditor.onResponseBodyChange(JSON.stringify({
          newBody: 'true'
      }));
    let tree = endpointEditor.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('onResponseBodyChange validates JSON before updating state', () => {
    endpointEditor.root
      .findByProps({ hasEditor: true })
      .props.endpointEditor.onResponseBodyChange("{ 'incomplete': 'JSON");
    let tree = endpointEditor.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
