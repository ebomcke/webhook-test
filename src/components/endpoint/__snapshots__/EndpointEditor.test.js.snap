// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interactions tests onResponseBodyChange updates state value 1`] = `
<div
  endpointEditor={
    Object {
      "endpoint": Object {
        "defaultResponseBody": "{\\"newBody\\":\\"true\\"}",
        "defaultResponseStatusCode": 200,
        "id": undefined,
        "path": "",
      },
      "onResponseBodyChange": [Function],
      "onStatusCodeChange": [Function],
      "onUrlChange": [Function],
      "save": [Function],
    }
  }
  hasEditor={true}
/>
`;

exports[`Interactions tests onResponseBodyChange validates JSON before updating state 1`] = `
<div
  endpointEditor={
    Object {
      "endpoint": Object {
        "defaultResponseBody": "{
  \\"status\\": \\"OK\\"
}",
        "defaultResponseStatusCode": 200,
        "id": undefined,
        "path": "",
      },
      "onResponseBodyChange": [Function],
      "onStatusCodeChange": [Function],
      "onUrlChange": [Function],
      "save": [Function],
    }
  }
  hasEditor={true}
/>
`;

exports[`Interactions tests onStatusCodeChange updates state value 1`] = `
<div
  endpointEditor={
    Object {
      "endpoint": Object {
        "defaultResponseBody": "{
  \\"status\\": \\"OK\\"
}",
        "defaultResponseStatusCode": 418,
        "id": undefined,
        "path": "",
      },
      "onResponseBodyChange": [Function],
      "onStatusCodeChange": [Function],
      "onUrlChange": [Function],
      "save": [Function],
    }
  }
  hasEditor={true}
/>
`;

exports[`Interactions tests onUrlChange updates state value 1`] = `
<div
  endpointEditor={
    Object {
      "endpoint": Object {
        "defaultResponseBody": "{
  \\"status\\": \\"OK\\"
}",
        "defaultResponseStatusCode": 200,
        "id": undefined,
        "path": "new-url-value",
      },
      "onResponseBodyChange": [Function],
      "onStatusCodeChange": [Function],
      "onUrlChange": [Function],
      "save": [Function],
    }
  }
  hasEditor={true}
/>
`;

exports[`Interactions tests onUrlChange validates url before updating state 1`] = `
<div
  endpointEditor={
    Object {
      "endpoint": Object {
        "defaultResponseBody": "{
  \\"status\\": \\"OK\\"
}",
        "defaultResponseStatusCode": 200,
        "id": undefined,
        "path": "newurlvalue",
      },
      "onResponseBodyChange": [Function],
      "onStatusCodeChange": [Function],
      "onUrlChange": [Function],
      "save": [Function],
    }
  }
  hasEditor={true}
/>
`;

exports[`Rendering tests Existing endpoint 1`] = `
<div
  endpointEditor={
    Object {
      "endpoint": Object {
        "defaultResponseBody": "{\\"unitTest\\":\\"true\\"}",
        "defaultResponseStatusCode": 418,
        "id": "existing-endpoint",
        "path": "existing-endpoint-path",
      },
      "onResponseBodyChange": [Function],
      "onStatusCodeChange": [Function],
      "onUrlChange": [Function],
      "save": [Function],
    }
  }
/>
`;

exports[`Rendering tests New endpoint 1`] = `
<div
  endpointEditor={
    Object {
      "endpoint": Object {
        "defaultResponseBody": "{
  \\"status\\": \\"OK\\"
}",
        "defaultResponseStatusCode": 200,
        "id": undefined,
        "path": "",
      },
      "onResponseBodyChange": [Function],
      "onStatusCodeChange": [Function],
      "onUrlChange": [Function],
      "save": [Function],
    }
  }
/>
`;
